2.0-PROPOSAL/README.md

```diff
@@ -0,0 +1,885 @@
+isolating Brigade's overall awareness of Kubernetes as the underlying workload
+execution substrate to only a few components.
+
+### High Level Architecture
+
+With a few cross-cutting concerns such as access control notwithstanding, the
+proposed architecture decomposes Brigade into three logical subsystems-- record
+storage, scheduling, and logging-- with a service layer to coordinate among
+these three. The service layer will be exposed to clients via secure HTTP
+endpoints that implement a RESTful API. Brigade types and an API client will
+also be made available in a Go SDK. An improved `brig` CLI and new gateways will
+utilize that SDK.
+
+![Logical Component Model](components.png)
+
+N.B.: Brigade 1.x has a read-only API that is utilized by Kashti only. (Kashti
+is the web-based, read-only Brigade 1.x dashboard.) The API referenced above is
+a _new_ one for use by all Brigade 2.0 components, gateways, and clients so as
+to abstract all of those consumers away from underlying technology choices.
+
+N.B.: A Brigade 2.x-compatible Kashti 2.0 is not proposed at this time.
+
+The following diagram depicts by example how different logical components
+interact with one another, using the complex sequence of event creation and
+handling to illustrate.
+
+![Event Lifecycle](events.png)
+
+### API Endpoints
+
+Brigade 2.0's REST API endpoints will be thin, serving only to handle
+transport-specific details-- decoding HTTP requests into domain objects that may
+be passed to the transport-agnostic service layer and encoding domain objects
+returned from that service layer as HTTP responses.
+
+It is proposed that validation be handled by the API endpoints since JSON schema
+provides an easy and efficient mechanism for this purpose, but must be applied
+to raw HTTP request bodies prior to unmarshaling.
+
+N.B.: In the case of projects, JSON schema can be reused by IDEs to validate
+project definitions as they are edited. The author proposes this UX to be
+superior to the current `brig project create` process.
+
+### Service Layer
+
+Brigade 2.0's service layer will implement high level, vendor neutral business
+logic and coordinate Brigade's underlying subsystems through well-defined
+interfaces.
+
+### Record Storage
+
+The record storage subsystem will provide a synchronous interface for the
+persistence of domain objects such as projects and events. Document-based
+storage is proposed since relatively few relationships exist between domain
+objects and among those relationships that do exist, composition is highly
+favored. i.e. A relational database is unlikely to yield any significant
+advantages and would also introduce object-relational impedance that can
+otherwise be avoided.
+
+Further, after careful consideration of MongoDB, Cassandra, and CouchDB, as well
+as cursory consideration of a few other options, MongoDB is proposed as the
+underlying data store based on several factors:
```

Comment:
From @brandonkal:

> Have you considered Postgres?

With finite bandwidth and a very large number of contenders in this field, summary dismissal of some candidates was necessary in order to allow closer examination of others that were not as easily dismissed. (There was a primary, if you will.)

From the start, it was thought that document-based storage would offer the least impedance with the domain model. Brigade 1.x can be cited as historical proof of this, for storing project and event data in Kubernetes secrets is tantamount to a makeshift document store.

I presume you're not suggesting using Postgres in a relational capacity (please correct me if I am wrong) and are instead referring to its support for storing JSON documents.

Again, having being necessary to dismiss some candidates so that more promising ones might receive closer scrutiny, Posgtres was admittedly dismissed on the basis that document storage seemed an afterthough-- something "tacked on" to their existing product.

You're the second person to suggest a closer look at Postgres (with the other suggestion having come offline from a former colleague). I'm willing to give it further scrutiny.

Reply:
Yes, I am referring to it as a JSON document store. Some useful features:
1. RLS. The database itself can determine if a tenant has access to a particular document.
2. It's simple and less Node.js centric
3. You can run [TRIGGERS](https://www.postgresql.org/docs/current/trigger-example.html) (i.e. written with JS or another language)
4. It can optionally act as a [message broker](https://github.com/andywer/pg-listen) which is inline with the simplicity goals here.

Reply:
@brandonkal after reviewing Postgres' JSON support, I am still of the opinion that MongoDB is a better fit. I only got as far as writing some relatively simple queries because I encountered a problematic detail early on-- Postgres isn't schema-less. i.e. You still need to create tables to hold your JSON blobs. That's fine, except that in our case that would introduce a new complication in terms of how we would automate database migrations at the time of Helm chart install/upgrade-- which is a thing that last I attempted it was doable, but not without its own complications.

I'd prefer to avoid unnecessary complications at this juncture, so my initial recommendation of MongoDB stands.

However, I _am_ updating the proposal to be much more explicit about loose coupling to the underlying data store to preserve the possibility of support for alternative data stores being contributed by the community in the future.